### ISSUE TITLE
ActivityPub Implementation

### ISSUE THREAD

**Issue by viveret** [pink-umbrella/pink-umbrella#42]
https://www.w3.org/TR/activitypub/
https://www.w3.org/ns/activitystreams

Main controllers: ActivityPubController

Motivations: https://www.w3.org/TR/activitystreams-vocabulary/#motivations
 - [x] Reactions

Deferred to V2:
 - [ ] Content / Collection / Relationship Management
 - [ ] Event RSVP / Geo-Social Events
 - [ ] Group Management
 - [ ] Questions / Answers
 - [ ] Notifications

Service Types:
- [x] Activity type listeners
- [x] Inbox / outbox / activity collection listeners
- [x] Data providers

**Comment by viveret** [comment]
Activities:
- [ ] Accept
- [ ] Add
- [ ] Announce
- [ ] Arrive
- [ ] Block
- [ ] Create
- [ ] Delete
- [ ] Dislike
- [ ] Flag
- [ ] Follow
- [ ] Ignore
- [ ] Invite
- [ ] Join
- [ ] Leave
- [ ] Like
- [ ] Listen
- [ ] Move
- [ ] Offer
- [ ] Question
- [ ] Reject
- [ ] Read
- [ ] Remove
- [ ] TentativeReject
- [ ] TentativeAccept
- [ ] Travel
- [ ] Undo
- [ ] Update
- [ ] View

**Comment by viveret** [comment]
Actors https://www.w3.org/TR/activitystreams-vocabulary/#actor-types:
- [ ]    Application
- [ ]    Group
- [ ]    Organization
- [ ]    Person
- [ ]    Service

**Comment by viveret** [comment]
Object and Link Types https://www.w3.org/TR/activitystreams-vocabulary/#object-types:
- [ ]    Article
- [ ]    Audio
- [ ]    Document
- [ ]    Event
- [ ]    Image
- [ ]    Note
- [ ]    Page
- [ ]    Place
- [ ]    Profile
- [ ]    Relationship
- [ ]    Tombstone
- [ ]    Video

**Comment by viveret** [comment]
Core types https://www.w3.org/TR/activitystreams-vocabulary/#types:
- [ ]    Object
- [ ]    Link
- [ ]    Activity
- [ ]    IntransitiveActivity
- [ ]    Collection
- [ ]    OrderedCollection
- [ ]    CollectionPage
- [ ]    OrderedCollectionPage

**Comment by sainteos** [comment]
I started this project a while back: https://gitlab.com/hexiea_feint/bustle_pub Though, I am unclear what our topology will look like, so we might want it in C#

**Comment by viveret** [comment]
@sainteos it's good you have experience in this, I'm going to implement it in C# so that the client and server can use activitypub / activitystreams as the data model. I can easily map what I've done so far in SQL to this, and clean up the SQL so that we can use activitystream datastores as a decentralized model

**Comment by viveret** [comment]
Data model completed, having read through the entire spec it is really nice and will fit most of our needs. We can also add our own types and extend the spec. 733bb73

**Comment by viveret** [comment]
Error codes
- [ ] Servers MUST return a 201 Created HTTP on success for an action (and URL to view it)
- [ ] Method not allowed
- [ ] Rate limiting

**Comment by viveret** [comment]
Depends on #10
